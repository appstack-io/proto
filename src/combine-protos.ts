import * as fs from 'fs';
import * as path from 'path';
import { consolidateProtos } from './consolidate-protos';

const findProtoFiles = (
  dir: string,
  exclude: string[],
  fileList: string[] = [],
): string[] => {
  const files: string[] = fs.readdirSync(dir);

  files.forEach((file) => {
    const filePath: string = path.join(dir, file);
    const stat: fs.Stats = fs.statSync(filePath);

    if (stat.isDirectory()) {
      findProtoFiles(filePath, exclude, fileList);
    } else if (path.extname(file) === '.proto') {
      fileList.push(filePath);
    }
  });

  return filterStrings(fileList, exclude);
};

function filterStrings(strings: string[], exclude: string[]) {
  return strings.filter(
    (str: string) => !exclude.some((excl) => str.includes(excl)),
  );
}

export function combineProtos(sources: string[], exclude: string[]) {
  const protoFiles: string[] = sources.reduce((acc, val) => {
    acc.push(...findProtoFiles(val, exclude));
    return acc;
  }, []);

  let combinedContent = '';
  const importStatements: Set<string> = new Set();

  // Loop through all files
  protoFiles.forEach((file) => {
    const content: string = fs.readFileSync(file, 'utf8');

    // Extract unique import statements
    content.split('\n').forEach((line) => {
      if (line.startsWith('import ')) {
        importStatements.add(line);
      }
    });

    // Exclude lines with 'syntax = "proto3";', 'import ...', or starting with 'package'
    const filteredContent: string = content
      .split('\n')
      .filter(
        (line) =>
          !line.includes('syntax = "proto3";') &&
          !line.startsWith('import ') &&
          !line.startsWith('package '),
      )
      .join('\n');

    combinedContent += filteredContent + '\n';
  });

  // Consolidate the combined content
  const consolidated: string = consolidateProtos(combinedContent);

  // Prepend 'syntax = "proto3";' and the unique import statements to the consolidated content
  const finalContent: string =
    'syntax = "proto3";\n' +
    Array.from(importStatements).join('\n') +
    '\npackage main;\n\n' +
    consolidated;

  return finalContent;
}
