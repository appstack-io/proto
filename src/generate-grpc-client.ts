import { exec } from 'child_process';
import * as fs from 'fs';

export async function generateGrpcClient(
  nodeModulesDir: string,
  workDir: string,
  sourceProto: string,
  targetTs: string,
) {
  await new Promise<void>((resolve, reject) => {
    exec(
      `${nodeModulesDir}/.bin/grpc_tools_node_protoc --plugin=protoc-gen-ts_proto=node_modules/.bin/protoc-gen-ts_proto --ts_proto_out=./${workDir} --ts_proto_opt=outputServices=nice-grpc,outputServices=generic-definitions,useExactTypes=false --proto_path=. ${workDir}/${sourceProto}`,
      (error, stdout, stderr) => {
        if (error || stderr) {
          reject(error || stderr);
        } else {
          resolve();
        }
      },
    );
  });
  fs.cpSync(
    `${workDir}/${workDir}/${sourceProto.replace('proto', 'ts')}`,
    `${workDir}/${targetTs}`,
  );
  fs.rmSync(`${workDir}/${workDir.split('/')[0]}`, {
    recursive: true,
  });
  const generated = fs.readFileSync(`${workDir}/${targetTs}`, 'utf8');
  const cleaned = generated.replace(/from ".*\/google/g, `from "./google/`);
  fs.writeFileSync(`${workDir}/${targetTs}`, cleaned);
}
