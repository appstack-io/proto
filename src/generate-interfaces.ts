export function generateInterfaces(source: string) {
  const interfaceRegex = /export (interface|enum) [\s\S]*?\}(?=\n|$)/gm;
  const imports: string[] = [
    "import { Empty } from './google/protobuf/empty';",
    "import { Observable } from 'rxjs';",
  ];
  const interfaces: string[] | null = source.match(interfaceRegex);
  return [...imports, ...(interfaces || [])].join('\n\n');
}
