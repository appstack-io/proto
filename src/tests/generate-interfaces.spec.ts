import { combineProtos } from '../combine-protos';
import { generateInterfaces } from '../generate-interfaces';
import * as fs from 'fs';
import { generateGrpcClient } from '../generate-grpc-client';

describe('generate-interfaces', () => {
  test('basic', async () => {
    combineProtos([`${__dirname}/protos`], []);
    await generateGrpcClient(
      `./node_modules`,
      `src/tests/temp`,
      `combined.proto`,
      `combined.grpc.client.ts`,
    );
    const client = fs.readFileSync(
      `${__dirname}/temp/combined.grpc.client.ts`,
      'utf-8',
    );
    const interfaces = generateInterfaces(client);
    expect(interfaces).toMatchSnapshot();
  });
});
