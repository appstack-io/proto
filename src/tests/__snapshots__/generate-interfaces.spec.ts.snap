// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-interfaces basic 1`] = `
"import { Empty } from './google/protobuf/empty';

import { Observable } from 'rxjs';

export interface Test1 {
  id: string;
  createdAt: number;
  updatedAt: number;
  text: string;
  isPublic: boolean;
}

export interface Test1CreateOneInput {
  text: string;
}

export interface Test2 {
  id: string;
  createdAt: number;
  updatedAt: number;
  text: string;
  isPublic: boolean;
}

export interface Test2CreateOneInput {
  text: string;
}

export interface Test1ServiceImplementation<CallContextExt = {}> {
  createOne(request: Test1CreateOneInput, context: CallContext & CallContextExt): Promise<DeepPartial<Test1>>;
}

export interface Test1ServiceClient<CallOptionsExt = {}> {
  createOne(request: DeepPartial<Test1CreateOneInput>, options?: CallOptions & CallOptionsExt): Promise<Test1>;
}

export interface Test2ServiceImplementation<CallContextExt = {}> {
  createOne(request: Test2CreateOneInput, context: CallContext & CallContextExt): Promise<DeepPartial<Test2>>;
}

export interface Test2ServiceClient<CallOptionsExt = {}> {
  createOne(request: DeepPartial<Test2CreateOneInput>, options?: CallOptions & CallOptionsExt): Promise<Test2>;
}"
`;
